Adding static route and ECN changes needed to pass.

rraghav-cisco opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
For cisco-8000, if the traffic flows to a connected route, the system uses the default voq. If we use static route, it uses the split-voq. In this PR, we change the dest IP address of the traffic streams to a static routed destination, and add that static route as well in the DUT. To test it, we have used the ECN script, and had to modify it as well.

In ECN we keep hitting the drop or other limits before hitting the xoff limits, due to the default voq. So changing it to static route helps in passing the tests.
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405
- [X] 202411

### Approach
#### What is the motivation for this PR?
Hitting failures in multiple tests due to traffic taking default voq (due to the connected routes).

#### How did you do it?
Added a new function to add static route in DUT, and use the new IP destination value for the traffic stream.

#### How did you verify/test it?
Alpesh had run the code in 202405 in my TB.

#### Any platform specific information?
Only for cisco-8000.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

sdszhang commented:
@rraghav-cisco can you resolve the conflict? and is the log analyzer failure still there, or should we add  a ignore for that log analyzer failure?

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

sdszhang commented on tests/common/snappi_tests/snappi_fixtures.py:
```diff
@@ -857,6 +858,9 @@ def __intf_config_multidut(config, port_config_list, duthost, snappi_ports, setu
         else:
             cmd = "remove"
+        if not setup:
+            static_routes_cisco_8000(tgenIp, duthost, port['peer_port'], port['asic_value'], setup)
```
instead of configuring static route for all test cases, can we use a fixture to configure it? then we can add that fixture for the test case which needs this.

rraghav-cisco replied:
Introducing the new function as a fixture was my first choice. But it was not working out, because we don't know the dest IP address until the streams are configured. That is why I picked this approach.  I am happy to check back on the fixture approach.

rraghav-cisco commented on tests/snappi_tests/ecn/test_ecn_marking_with_snappi.py:
```diff
@@ -109,9 +158,22 @@ def test_ecn_marking_port_toggle(
 
 
-test_flow_percent_list = [[90, 15], [53, 49], [15, 90], [49, 49], [50, 50], [60, 60], [60, 90], [90, 60]]
+test_flow_percent_list = [
```
This change was added since more combinations of TC3 and TC4 were needed to be tested. I will remove this from this PR, and introduce it in a different PR for ECN.

rraghav-cisco commented on tests/common/snappi_tests/snappi_fixtures.py:
```diff
@@ -1292,3 +1298,57 @@ def check_fabric_counters(duthost):
                               format(fec_uncor_err, duthost.hostname, val_list[0], val_list[1]))
+
+
+MAPPING = {}
```
Yes, I agree. I will use DEST_TO_GATEWAY_MAP, if that is ok.

rraghav-cisco commented on tests/common/snappi_tests/snappi_fixtures.py:
```diff
@@ -1292,3 +1298,57 @@ def check_fabric_counters(duthost):
                               format(fec_uncor_err, duthost.hostname, val_list[0], val_list[1]))
+
+
+MAPPING = {}
```
Actually, since the mapping is static, we may even get rid of the mapping - and calculate the required destination on the fly. Lets discuss it in our meeting.

rraghav-cisco commented on tests/common/snappi_tests/snappi_fixtures.py:
```diff
@@ -1292,3 +1298,57 @@ def check_fabric_counters(duthost):
+            raise RuntimeError(f"Request for dest addr: {addr} without setting it in advance.")
+        return MAPPING[addr]['dest']
+
+    if dut.facts['asic_type'] != "cisco-8000":
```
We need it for the T0/T1 as well. I am going through the exercise of adding support for t0 and t1. The T1 is good, T0 seems to hit some failures, I am debugging them.

rraghav-cisco commented on tests/snappi_tests/ecn/test_ecn_marking_with_snappi.py:
```diff
@@ -109,9 +158,22 @@ def test_ecn_marking_port_toggle(
+    [52, 50]]
 
 
+@pytest.mark.parametrize("number_of_streams", [1, 10])
```
The ECN tests could be tested with single or multiple streams. This will be removed in this PR.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

rraghav-cisco commented:
@sdszhang , I have addressed the comments discussed in the meeting. 
1. Restrict the new code to only modular chassis.
2. Add the routecheck to the loganalyzer skip messages.
3. Removed the extra paramters for the ECN case.
Pls review and let me know if you comments. I will attach logs as soon as I get a run. Thanks.

sdszhang commented:
> @sdszhang , I have addressed the comments discussed in the meeting.
> 
> 1. Restrict the new code to only modular chassis.
> 2. Add the routecheck to the loganalyzer skip messages.
> 3. Removed the extra paramters for the ECN case.
>    Pls review and let me know if you comments. I will attach logs as soon as I get a run. Thanks.

lgtm.  Thanks for the quick turnaround.

mssonicbld commented:
/azp run

rraghav-cisco commented:
Finally got first ECN case to pass:
`=========================================================================================================== PASSES ===========================================================================================================
__________________________________________________________________________ test_ecn_marking_lossless_prio[multidut_port_info0-test_flow_percent0-1] __________________________________________________________________________
__________________________________________________________________________ test_ecn_marking_lossless_prio[multidut_port_info1-test_flow_percent0-1] __________________________________________________________________________
--------------------------------------------------------------- generated xml file: /run_logs/ixia/ecn-static-1/2025-03-12-01-15-57/tr_2025-03-12-01-15-57.xml ---------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
================================================================================================== short test summary info ===================================================================================================
PASSED snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py::test_ecn_marking_lossless_prio[multidut_port_info0-test_flow_percent0-1]
PASSED snappi_tests/multidut/ecn/test_multidut_ecn_marking_with_snappi.py::test_ecn_marking_lossless_prio[multidut_port_info1-test_flow_percent0-1]
SKIPPED [1] common/helpers/assertions.py:22: Egress port must be on short link
=================================================================================== 2 passed, 1 skipped, 6 warnings in 1217.97s (0:20:17) ====================================================================================
sonic@snappi-sonic-mgmt-vanilla-202405-t2:/data/tests$ check
`

Running the full ecn folder now.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
Cherry-pick PR to 202411: https://github.com/sonic-net/sonic-mgmt/pull/17592

amitpawar12 commented:
This merge is causing existing T2 testcases to break. Issue #17611 raised for the same.

yejianquan commented:
Hi @rraghav-cisco  @sdszhang , is this PR in msft-202405 already? If so, please remove the conflict tag and add included tag

sdszhang commented:
@yejianquan This was manually cherry-picked into msft-202405 in https://github.com/Azure/sonic-mgmt.msft/pull/159.  labels are updated now
